### ISSUE TITLE
new "path" interpolation variable for relative path output

### ISSUE THREAD

**Issue by thechile** [hashicorp/terraform#20064]
**Overview**

At the moment with our remote state being stored in S3, for certain resources that deal with filesystem paths e.g. binary files or archive files... the absolute path gets stored in the state file. Which means on subsequent runs by other people on different workstations terraform will wrongly say the file has changed due to the local path.. not the actual file having changed. For small files perhaps people can overlook this and will push the change every time but for larger files this can be a pain.  

**Example problem**

Using this code

`source = "${path.module}/binary_files/${data.null_data_source.vc_dist.outputs.name}"`

which resolves `source` to this type of thing: `/home/somebody/dev/tf/my-omdule/.terraform/modules/c686a49affc14555e3a4147bafd109fe/binary_files/vc_redist.x64_2015.zip`

when in fact in state file we want just

`.terraform/modules/c686a49affc14555e3a4147bafd109fe/binary_files/vc_redist.x64_2015.zip`

**Solution**

There are a few workarounds being used to avoid this scenario, the more common seem to be these examples:

_Example 1)_ 
`source = "${substr("${path.module}/binary_files/${data.null_data_source.vc_dist.outputs.name}", length(path.cwd) + 1, -1)}"`

_Example 2)_
`source = ".${replace(path.module, path.root, "")}/binary_files/${data.null_data_source.vc_dist.outputs.name}"`

Which just delete the absolute part of the path so everything is accessed via the `.terraform` symlinks to the respective directories.. ironically which are via the absolute path. But at least this way the state file is consistent and doesn't require `ignore_changes ` to be used. 

To remove the need of using `substr` or `replace` could it be helpful to introduce a new relative option for the interpolation path variable ? e.g. `path.relative_root` so the examples above just become:

`source = ".${path.relative_root}/binary_files/${data.null_data_source.vc_dist.outputs.name}"`

perhaps `path.root.relative` or `path.rroot` and `path.rmodule` are preferred. 

Sound OK ?

thanks

**Comment by apparentlymart** [comment]
Hi @thechile! Thanks for this feature request.

We've actually already changed the behaviors of `path.root` and `path.module` in master, ready for the forthcoming v0.12.0 release, to return relative paths. This means there will be one more errant change detected after upgrade (for those _not_ using `replace` as you are) but after that the paths should be portable between systems.

We are also changing the path variables and path functions to always use forward slashes, even on Windows, for similar reasons.

In practice we could find little reason to use absolute paths in `path.root` and `path.module`, but for those who _do_ need it we have retained `path.cwd` as an absolute path that can have `path.root` or `path.module` concatenated to it.

I think this will actually have an interesting, problematic consequence for your workaround: in most cases, `path.root` will simply be `.` representing the current directory, which would cause any dots in `path.module` -- including the one in `.terraform` -- to be stripped out. :confounded: I can't think of anything reasonable we could have the upgrade tool do for that situation, so this is something you'll need to watch out for on upgrade and make sure to undo workarounds like this for correct results.
